File: app/src/main/java/me/magnum/melonds/model/ControllerConfiguration.java
package me.magnum.melonds.model;

import java.util.ArrayList;
import java.util.List;

public class ControllerConfiguration {
	private static final Input[] configurableInput = {
			Input.A,
			Input.B,
			Input.X,
			Input.Y,
			Input.LEFT,
			Input.RIGHT,
			Input.UP,
			Input.DOWN,
			Input.L,
			Input.R,
			Input.START,
			Input.SELECT,
			Input.HINGE,
			Input.PAUSE,
			Input.FAST_FORWARD
	};

	public static ControllerConfiguration empty() {
		ArrayList<InputConfig> inputConfigs = new ArrayList<>();
		for (Input input : configurableInput) {
			inputConfigs.add(new InputConfig(input, -1));
		}

		return new ControllerConfiguration(inputConfigs);
	}

	private List<InputConfig> inputMapper;

	public ControllerConfiguration(List<InputConfig> inputMapper) {
		ArrayList<InputConfig> actualConfig = new ArrayList<>();

		for (Input input : configurableInput) {
			InputConfig inputConfig = null;
			for (InputConfig config : inputMapper) {
				if (config.getInput() == input) {
					inputConfig = config;
					break;
				}
			}

			if (inputConfig != null)
				actualConfig.add(inputConfig);
			else
				actualConfig.add(new InputConfig(input));
		}

		this.inputMapper = actualConfig;
	}

	public Input keyToInput(int key) {
		for (int i = 0; i < inputMapper.size(); i++) {
			if (inputMapper.get(i).getKey() == key)
				return inputMapper.get(i).getInput();
		}

		return null;
	}

	public List<InputConfig> getConfigList() {
		return this.inputMapper;
	}
}

File: app/src/main/java/me/magnum/melonds/domain/model/ControllerConfiguration.kt
package me.magnum.melonds.domain.model

class ControllerConfiguration(configList: List<InputConfig>) {
    companion object {
        private val configurableInput = listOf(
            Input.A,
            Input.B,
            Input.X,
            Input.Y,
            Input.LEFT,
            Input.RIGHT,
            Input.UP,
            Input.DOWN,
            Input.L,
            Input.R,
            Input.START,
            Input.SELECT,
            Input.HINGE,
            Input.PAUSE,
            Input.FAST_FORWARD,
            Input.MICROPHONE,
            Input.RESET,
            Input.SWAP_SCREENS,
            Input.QUICK_SAVE,
            Input.QUICK_LOAD,
            Input.REWIND,
            Input.REFRESH_EXTERNAL_SCREEN
        )
    }

    val inputMapper: List<InputConfig>

    init {
        inputMapper = configurableInput.map { input ->
            // Find config in the config list, or fallback to a default config without assignments
            configList.firstOrNull { it.input == input } ?: InputConfig(input)
        }
    }

    fun getInputAssignments(input: Input): List<InputConfig.Assignment>? {
        return inputMapper.firstOrNull { it.input == input && it.hasKeyAssigned() }?.let {
            listOfNotNull(
                it.assignment.takeIf { it != InputConfig.Assignment.None },
                it.altAssignment.takeIf { it != InputConfig.Assignment.None },
            )
        }
    }

    fun keyToInput(key: Int): Input? {
        for (config in inputMapper) {
            val assignments = listOf(config.assignment, config.altAssignment)
            if (assignments.any { (it as? InputConfig.Assignment.Key)?.keyCode == key }) {
                return config.input
            }
        }
        return null
    }

    fun axisToInput(axis: Int, direction: InputConfig.Assignment.Axis.Direction): Input? {
        return inputMapper.firstOrNull { config ->
            val assignments = listOf(config.assignment, config.altAssignment)
            assignments.any {
                (it as? InputConfig.Assignment.Axis)?.let { ax ->
                    ax.axisCode == axis && ax.direction == direction
                } ?: false
            }
        }?.input
    }
}